### DISCUSSION TITLE
Implement stable links to pages and headings

### DISCUSSION THREAD

**Discussion by addisonj** [markdoc/markdoc#239] [Q&A] [answered]
With markdoc, I am looking to have editors, especially in the early stages, often re-organizing and changing the structure of the content.
This will often mean that links will change both to pages but as well as to headings, with those headings potentially coming from a partial.

I am looking to implement a link that doesn't rely on the structure of the URLs but instead could use a stable ID to a page or heading.

It isn't immediately clear to me what would be the best way to implement this. I don't think an AST transformation alone would work, as I would need an AST for all the docs, not just the current document. 

Is there any prior art on how to go about building such a mechanism?

**Comment by nkohari-stripe** [comment, marked as answer]
There isn't really a Markdoc-specific way to do this, but I can suggest one approach. This assumes that you're using React, but it'd work with any other rendering mechanism.

First, add the stable identifier to the frontmatter of your documents, like:

```
id: my-awesome-article
---

Here's the article content
```

Second, create a custom component which overrides `link` nodes. For example:

```typescript
{
  node: 'link',
  render: 'Link',
  attributes: {
    href: { type: String },
  },
};
```

Third, gather the identifiers from your documents. This is the most complex piece, and is dependent on your build system. Broadly speaking, what you're looking for is a way to write document metadata to a module which can be imported in your `Link` component. There's an example of [one approach](https://github.com/nkohari/nate.io/tree/master/build) using Vite on my personal website, but I'm sure it'd be feasible to implement something like this using Next.js or something similar.

```typescript
const getUrlForDocumentWithId = (id: string) => {
  // Use the document metadata module that you created during your build step
  // to find the document with the specified id, and return its URL.
};

type LinkProps = React.AnchorHTMLAttributes<HTMLAnchorElement>;

const Link = ({href, ...props}: LinkProps) => {
  let url;
  if (href.startsWith('id:')) {
    const id = href.replace('id:', '');
    url = getUrlForDocumentWithId(id);
  } else {
    url = href;
  }

  return <a href={href} {...props} />;
}
```

Then, in your articles, you could create links using these identifiers like this. (Using the `id:` scheme in the URL is an arbitrary choice; you could do this however you'd prefer as long as your `Link` component understands it.)

```
Here is an [awesome article](id:my-awesome-article), you should read it!
```

Sorry, I know this is kind of hand-wavy around the hard part, but that's mostly because it's very implementation-dependent! You'll want something along these lines, though.

**Reply by addisonj** [reply]
Thanks! I appreciate the effort in this answer!

This was the direction I was thinking I likely would need to go, which is totally reasonable. I do wonder if markdoc (or likely a companion library for markdoc :thinking:) might benefit from a sort of library of documents and metadata. 

I am currently starting with a next.js site, but I expect to have a fairly large set of partials, if my work heads that direction I will try and share learnings/code here :)

**Reply by duncanma-stripe** [reply]
I was considering building something like "might benefit from a sort of library of documents and metadata" into my own build script, so that I could create list pages, RSS, etc. It seems like something external to markdoc itself though
